### ISSUE TITLE
Spec failures on jruby

### ISSUE THREAD

**Issue by fran-worley** [dry-rb/dry-validation#159]
2 tests in ./spec/integration/schema/predicates/lt_spec.rb:154 and :23 are failing with the same result:

#Failure/Error: expected that #<Dry::Validation::Result output={} messages={:foo=>["is missing"]}> would be failing (["is missing", "must be less than  23"])

I've temporarily made them pending to keep the build green but can we either make jruby and accepted failure or work out why they fail some of the time on jruby?

It doesn't happen every time, and it doesn't matter the seed order and it only occurs in jruby 9000.

**Comment by flash-gordon** [comment]
@fran-worley you can [see here](https://travis-ci.org/dry-rb/dry-validation/jobs/129469179) that travis actually uses 9.0.0.0.pre1, not released version. That's how I handled it in dry-container: https://github.com/dry-rb/dry-container/pull/18/commits/a6ad6bace74e4e4536e05d08389f24e1a819d427 and [this comment](https://github.com/dry-rb/dry-container/pull/18#issuecomment-215986890) provides more context on the issue. TL;DR RVM need to be updated.

**Comment by fran-worley** [comment]
@flash-gordon Thanks for this. I've just run the spec against jruby 9.0.0.0 locally and the errors are still there so I don't think that changing travis to use the full version is going to be enough to sort this one.

**Comment by fran-worley** [comment]
These are the exact errors:

```ruby
Failures:

  1) Predicates: Lt with required with missing input is not successful
     Failure/Error: expect(result).to be_failing ['is missing', 'must be less than 23']
       expected that #<Dry::Validation::Result output={} messages={:foo=>["is missing"]}> would be failing (["is missing", "must be less than 23"])
     # ./spec/integration/schema/predicates/lt_spec.rb:22:in `block in (root)'

  2) Predicates: Lt as macro with required with value with missing input is not successful
     Failure/Error: expect(result).to be_failing ['is missing', 'must be less than 23']
       expected that #<Dry::Validation::Result output={} messages={:foo=>["is missing"]}> would be failing (["is missing", "must be less than 23"])
     # ./spec/integration/schema/predicates/lt_spec.rb:152:in `block in (root)'
```

It looks like it is something to do with the hint compiler but I can't explain why it only seems to be a problem in these two specs.

**Comment by flash-gordon** [comment]
@fran-worley that's very interesting indeed! Although I'm not an expert in jruby I found a reproducible failing by brute forcing:
`rspec --seed 44023 ./spec/integration/schema/predicates/lt_spec.rb:23 ./spec/integration/hints_spec.rb:66` at jruby 9.1.0.0
Can't say anything more right now, it's 2AM here 🌙 :)

**Comment by fran-worley** [comment]
Thanks @flash-gordon I have zero jruby experience myself. Would it be best to make jRuby an acceptable fail in travis CI and take pending off the tests until someone can work it out?

**Comment by flash-gordon** [comment]
@fran-worley I don't have any preferences about that, feel free to comment specs for now. I'm going to fix the issue tomorrow, it's somehow related with class-level cache so I guess it won't be hard.

**Comment by flash-gordon** [comment]
```ruby
# hint_compiler.rb
      def call
        self.class.cache.fetch_or_store(hash) do # <----- the bug sits here :)
          super(rules)
        end
      end
```

**Comment by fran-worley** [comment]
Is it anything too do with this?

https://github.com/ruby-concurrency/concurrent-ruby/issues/512

Which inspired these docs?
- (Object) fetch_or_store(key, default_value = NULL)

The "fetch-then-act" methods of Map are not atomic. Map is intended to be use as a concurrency primitive with strong happens-before guarantees. It is not intended to be used as a high-level abstraction supporting complex operations. All read and write operations are thread safe, but no guarantees are made regarding race conditions between the fetch operation and yielding to the block. Additionally, this method does not support recursion. This is due to internal constraints that are very unlikely to change in the near future.

[View source<http://ruby-concurrency.github.io/concurrent-ruby/Concurrent/Map.html#>]


On 12 May 2016, at 00:16, Nikita Shilnikov <notifications@github.com<mailto:notifications@github.com>> wrote:


# hint_compiler.rb
      def call
        self.class.cache.fetch_or_store(hash) do # <----- the bug sits here :)
          super(rules)
        end
      end

-

**Comment by flash-gordon** [comment]
@fran-worley I raised an issue in jruby tracker: https://github.com/jruby/jruby/issues/3884 Will see what they think

**Comment by fran-worley** [comment]
Thanks @flash-gordon I'll have to leave this to someone better at the math/science than me. I'll leave the tests as pending for now.

**Comment by flash-gordon** [comment]
@fran-worley I excluded jruby-9000 from the run list. I will add 9.1.1.0 version instead when it is out (in a couple of days I guess). The issue was caused by incorrect hashing of deeply nested arrays of certain structure. In dry-v we use them for representing the ast (yeah, you know it :)) and hash them for caching. Invalid hash value leads to an incorrect cache hit which in order leads to a spec failure. Hash function was patched in 9.1.1.0 and will be completely rewritten in 9.1.2.0. As a side effect new version will use 64-bit hash values which greatly reduce probability of collisions.
